src/result_processor.c

```diff
@@ -627,294 +634,258 @@ static int rploaderNext(ResultProcessor *base, SearchResult *r) {
 
-  // Buffer management
-  SearchResult **BufferBlocks;
-  size_t BlockSize;
-  size_t buffer_results_count;
+  // get current block
+  SearchResult *curr_block = self->BufferBlocks[curr_elem_index / blockSize];
 
-  // Results iterator
-  size_t curr_result_index;
+  // get the result in the block
+  SearchResult* ret = curr_block + (curr_elem_index % blockSize);
 
-  // Redis's lock status
-  bool isRedisLocked;
+  // Increase result's index
+  ++self->curr_result_index;
 
-  // Last buffered result code. To know weather to return OK or EOF.
-  char last_buffered_rc;
-};
-/*********** Buffered and locker functions declarations ***********/
+  // return result
+  return ret;
+}
 
-// Destroy
-static void RPBufferAndLocker_Free(ResultProcessor *base);
+static int rpSafeLoaderNext_Accumulate(ResultProcessor *rp, SearchResult *res);  // Forward declaration
 
-// Buffer phase
-static int rpbufferNext_bufferDocs(ResultProcessor *rp, SearchResult *res);
+static int rpSafeLoader_ResetAndReturnLastCode(RPSafeLoader *self) {
 
-// Yield results phase functions
-static int rpbufferNext_Yield(ResultProcessor *rp, SearchResult *result_output);
-static int rpbufferNext_ValidateAndYield(ResultProcessor *rp, SearchResult *result_output);
-static bool isResultValid(const SearchResult *res);
-static int RPBufferAndLocker_ResetAndReturnLastCode(ResultProcessor *rp);
+  self->base_loader.base.Next = rpSafeLoaderNext_Accumulate; // Reset the next function, in case we are in cursor mode
+  self->buffer_results_count = 0;
+  self->curr_result_index = 0;
 
-// Redis lock management
-static bool isRedisLocked(RPBufferAndLocker *bufferAndLocker);
-static void LockRedis(RPBufferAndLocker *rpBufferAndLocker, RedisModuleCtx* redisCtx);
-static void UnLockRedis(RPBufferAndLocker *rpBufferAndLocker, RedisModuleCtx* redisCtx);
+  int rc = self->last_buffered_rc;
+  self->last_buffered_rc = RS_RESULT_OK;
+  return rc;
+}
 
-/*********** Buffered results blocks management functions declarations ***********/
-static SearchResult *GetResultsBlock(RPBufferAndLocker *rpPufferAndLocker, size_t bloc_idx);
+/*********************************************************************************/
 
-// Insert result to the buffer.
-//If @param CurrBlock is full we add a new block and return it, otherwise returns @param CurrBlock.
-static SearchResult *InsertResult(RPBufferAndLocker *rpPufferAndLocker, SearchResult *resToBuffer, SearchResult *CurrBlock);
-static bool IsBufferEmpty(RPBufferAndLocker *rpPufferAndLocker);
+static void rpSafeLoader_ValidateAndLoad(RPSafeLoader *self) {
+  SearchResult *curr_res;
 
-static SearchResult *GetNextResult(RPBufferAndLocker *rpPufferAndLocker);
-/*******************************************************************************/
+  // iterate the buffer.
+  // TODO: implement `GetNextResult` that gets the current block to save calculation time.
+  while ((curr_res = GetNextResult(self))) {
+    // Skip invalid results
+    if (!isResultValid(curr_res)) {
```

Comment:
TODO: what if a result is expired? We turn on a flag in it, we should not drop it in this case (not in the loading phase, because maybe it expired in a previous loader, and definitely not in the yielding phase)